### ISSUE TITLE
"to" conversion not working with QList

### ISSUE THREAD

**Issue by matpen** [ericniebler/range-v3#1425]
New user here, so sorry if I am missing something obvious. I am trying to compile the following simple test, and I notice that the version using `QList` (as opposed to `std::vector`) raises a `no match for operator|` error ([attached](https://github.com/ericniebler/range-v3/files/4159107/error.txt)):

```cpp
#include <QList>
#include <vector>
#include <range/v3/view/transform.hpp>
#include <range/v3/range/conversion.hpp>
using namespace ranges::v3;

int main()
{
#if 0
    std::vector<int> numbers{1, 2, 3};
    auto doubles = numbers
        | view::transform([](int item) { return 2 * item; })
        | to<std::vector<int>>;
#else
    QList<int> numbers{1, 2, 3};
    auto doubles = numbers
        | view::transform([](int item) { return 2 * item; })
        | to<QList<int>>;
#endif
}
```

I am aware that there are better ways to achieve the same result, this is just a stripped-down example. Just wondering if it is a bug in `range-v3` or some incompatibility in `QList` (which actually claims to be stl-compatible).

Using:
* `Qt 5.12.0`
* `range-v3 0.10.0`
* `gcc 7.4.0`

**Comment by vpicaver** [comment]
This is because QList in 5.12.0 doesn't support range constructor. It was added Qt 5.14. This is also support with QVector. Here's an example of the constructor that you need: 

QVector(InputIterator first, InputIterator last)

**Comment by matpen** [comment]
This is most insightful, thanks @vpicaver for the hint. I will try and upgrade my Qt installation as soon as possible, then!

**Comment by ericniebler** [comment]
@matpen, can you report back here if newer Qt versions fix this issue?

**Comment by matpen** [comment]
@ericniebler I just had the chance to try with the following new environment, which I setup from scratch just for the test:
* `Qt 5.14.0`
* `range-v3 0.10.0`
* `gcc 7.5.0`

But I am getting the very same error as descried in the OP (the new error log is [attached](https://github.com/ericniebler/range-v3/files/4398869/error.txt), you can diff the two if you wish).

I can confirm that the new version has the range constructor mentioned by @vpicaver, but this seems not to solve the issue:
```
# grep InputIterator /opt/qt5/include/QtCore/qlist.h
    template <typename InputIterator, QtPrivate::IfIsInputIterator<InputIterator> = true>
    QList(InputIterator first, InputIterator last);
[...]
```

**Comment by BRevzin** [comment]
`QList` does not have an `allocator_type`, which `ranges::to` requires: https://github.com/ericniebler/range-v3/blob/master/include/range/v3/range/conversion.hpp#L250-L258

(why, @ericniebler?)

**Comment by matpen** [comment]
Just another hypothesis here, but I noticed errors like the following:
```
/opt/qt5/include/QtCore/qflags.h:173:50: note:   no known conversion for argument 1 from 'ranges::transform_view<ranges::ref_view<QList<int> >, main()::<lambda(int)> >' to 'QFlags<QArrayData::AllocationOption>::enum_type {aka QArrayData::AllocationOption}'
 Q_DECL_CONSTEXPR inline QFlags<Flags::enum_type> operator|(Flags::enum_type f1, Flags::enum_type f2) noexcept \
                                                  ^
/opt/qt5/include/QtCore/qarraydata.h:126:1: note: in expansion of macro 'Q_DECLARE_OPERATORS_FOR_FLAGS'
 Q_DECLARE_OPERATORS_FOR_FLAGS(QArrayData::AllocationOptions)
```
Is it possible that there is a conflict between the pipe `|` operator used by ranges and by [QFlags](https://doc.qt.io/qt-5/qflags.html)?

**Comment by BRevzin** [comment]
I don't think so - one of the things that tends to happen when the function you _intend_ to call (in this case `to`) doesn't work is that the compiler goes ahead and gives you every other possible `|` it can find and tell you why that one doesn't work either. There's no conflict here - that one isn't viable.

**Comment by ericniebler** [comment]
> `QList` does not have an `allocator_type`, which `ranges::to` requires: [...]
>
> (why, @ericniebler?)

That's a holdover from when all views had implicit conversions to containers, and I needed a quick way to weed out not-container-ish things before getting too far down the path of instantiation the conversion operator. It's not needed anymore since I removed the implicit conversions operator.

Can one of you check to see if things work now?

**Comment by matpen** [comment]
@ericniebler Sorry for late reply. I just repeated the test with the latest master, and it seems to work correctly. Environment is same as last time, just with upgraded `range-v3`:

- Qt 5.14.0
- range-v3 @ commit 85abde093a3e
- gcc 7.5.0

Also, I performed the test with Qt 5.12.0, where I am getting the same [error](https://github.com/ericniebler/range-v3/files/4771204/error.log) as before. This might be expected, due to the missing range constructor mentioned [earlier in this thread](https://github.com/ericniebler/range-v3/issues/1425#issuecomment-594074860).

If that is the case, I think this issue can be considered solved (and I thank you for the great work!).
